Godeps/_workspace/src/k8s.io/kubernetes/pkg/proxy/userspace/unidlersocket.go

```diff
@@ -0,0 +1,365 @@
+	}
+	return nil, fmt.Errorf("unknown protocol %q", protocol)
+}
+
+// tcpUnidlerSocket implements proxySocket.  Close() is implemented by net.Listener.  When Close() is called,
+// no new connections are allowed but existing connections are left untouched.
+type tcpUnidlerSocket struct {
+	tcpProxySocket
+	eventRecorder record.EventRecorder
+}
+
+func (tcp *tcpUnidlerSocket) waitForEndpoints(ch chan<- interface{}, service proxy.ServicePortName, proxier *Proxier) {
+	defer close(ch)
+	for {
+		if proxier.loadBalancer.ServiceHasEndpoints(service) {
+			// we have endpoints now, so we're finished
+			return
+		}
+
+		// otherwise, wait a bit before checking for endpoints again
+		time.Sleep(endpointDialTimeout[0])
+	}
+}
+
+func (tcp *tcpUnidlerSocket) acceptConns(ch chan<- net.Conn, myInfo *serviceInfo) {
+	defer close(ch)
+
+	// Block until a connection is made.
+	for {
+		inConn, err := tcp.Accept()
+		if err != nil {
+			if isTooManyFDsError(err) {
+				panic("Accept failed: " + err.Error())
+			}
+
+			// TODO: indicate errors here?
+			if isClosedError(err) {
+				return
+			}
+			if !myInfo.isAlive() {
+				// Then the service port was just closed so the accept failure is to be expected.
+				return
+			}
+			glog.Errorf("Accept failed: %v", err)
+			continue
+		}
+
+		ch <- inConn
+	}
+}
+
+func (tcp *tcpUnidlerSocket) ProxyLoop(service proxy.ServicePortName, myInfo *serviceInfo, proxier *Proxier) {
+	if !myInfo.isAlive() {
+		// The service port was closed or replaced.
+		return
+	}
+
+	// accept connections asynchronously
+	inConns := make(chan net.Conn)
+	go tcp.acceptConns(inConns, myInfo)
+
+	endpointsAvail := make(chan interface{})
+	var allConns *connectionList
+
+	svcName := fmt.Sprintf("%s/%s:%s", service.Namespace, service.Name, service.Port)
+
+TOP_LOOP:
+	for {
+		glog.V(4).Infof("unidling TCP proxy start/reset for service %s/%s:%s", service.Namespace, service.Name, service.Port)
+		// collect connections and wait for endpoints to be available
+		sent_need_pods := false
+		timeout_started := false
+		ticker := time.NewTicker(needPodsTickLen)
+		allConns = newConnectionList(maxHeldConnections, needPodsTickLen, needPodsWaitTimeout, svcName)
+
+	WAIT_LOOP:
+		for {
+			select {
+			case inConn, ok := <-inConns:
+				if !ok {
+					// the listen socket has been closed, so we're finished accepting connections
+					break TOP_LOOP
+				}
+
+				if !sent_need_pods && !proxier.loadBalancer.ServiceHasEndpoints(service) {
+					// TODO: There's a constant in origin for this NeedPods
+					// HACK: make the message different to prevent event aggregation
+					glog.V(4).Infof("unidling TCP proxy sent unidle event to wake up service %s/%s:%s", service.Namespace, service.Name, service.Port)
+					tcp.eventRecorder.Eventf(myInfo.ref, api.EventTypeNormal, unidling.NeedPodsReason, "The service-port %s/%s:%s needs pods (at %v)!", service.Namespace, service.Name, service.Port, time.Now())
```

Comment:
> is the namespace needed in the message? is it not inferred from the namespace of the event itself?

Yep, removed

> why is ports enumerated? isn't the idea that a service has no endpoints and therefore needs pods (independent of a port)?

Sure, but independently, it could be useful to see that information in the event (If I'm a user or admin looking at what caused something to unidle, etc)